Rewrite the expression inference rules (#386)
ferrocene/specification: src/types-and-traits.rst

:t:`type inference root` being inferred. In the rules detailed below, a static
error occurs when :t:`type unification` fails.

:dp:`fls_6GrNr2izovRN`
Performing :t:`type inference` may introduce a requirement that some :t:`type`
must implement a :t:`trait`, or that a :t:`type` or :t:`lifetime` must outlive
some other :t:`lifetime`. Such requirements are referred to as
:dt:`[obligation]s` and are detailed in the inference rules below.

:dp:`fls_9dSltJ6U98Fo`
If insufficient :t:`type` information is available at the time an
:t:`obligation` is introduced, it may be deferred to be resolved later.
Any time new :t:`type` information is derived during :t:`type inference`, the

> tool shall attempt to resolve all outstanding :t:`[obligation]s` and propagate

Reviewer: ```suggestion
tool attempts to resolve all outstanding :t:`[obligation]s` and propagate
```
"shall" is for requirements aimed at the programmer.

Author: hmm, most of our "shall" requirements are actually checked by the tool it seems

     the contained :t:`block expression`, with the :t:`expected type` set to
     the :t:`never type`.

:dp:`fls_J6ydUCCJp1Sn`
:t:`Type inference` of :t:`[expression]s` may incorporate an :t:`expected type`,
derived from the context the :t:`expression` appears in. If the :t:`expression`
is a :t:`coercion site` or a :t:`coercion-propagating expression`, the :t:`type`
derived via :t:`type inference` may be coerced to the :t:`expected type`. If no
:t:`type coercion` to the :t:`expected type` is possible, or the :t:`expression`
is not a :t:`coercion site` or a :t:`coercion-propagating expression`, the
inferred :t:`expression` :t:`type` is unified with the :t:`expected type`.

:dp:`fls_FSQqHs8T4bUx`
The :t:`type inference` rules for :t:`[expression]s` are as follows:

* :dp:`fls_0HHC1iOk5dwz`
  An :t:`if expression` is inferred by inferring its :t:`subject expression`
  with an :t:`expected type` of :c:`bool`, then inferring its
  :t:`block expression` with the :t:`expected type` of the :t:`if expression`.
  Then, if the :t:`if expression` has an :t:`else expression`, apply the
  inference rules below to it.

* :dp:`fls_QZWTS0Giy3I3`

>   An :t:`if-let expression` is inferred by inferring its

Reviewer: ```suggestion
  An :t:`if let expression` is inferred by inferring its
```

  :t:`subject let expression` with the :t:`expected type` set to the :t:`type`
  of its :t:`pattern`, then inferring its :t:`block expression` with the
  :t:`expected type` of the :t:`if-let expression`. If the

>   :t:`if-let expression` has an :t:`else expression`, apply the inference rules

Reviewer: ```suggestion
  :t:`if let expression` has an :t:`else expression`, apply the inference rules
```

  below to it.

* :dp:`fls_KJsIu1lgVZxP`
  An :t:`else expression` that is part of an :t:`if expression` or

>   :t:`if-let expression` is inferred as follows:

Reviewer: ```suggestion
  :t:`if let expression` is inferred as follows:
```

  * :dp:`fls_KRQxdSav1KBA`
    If the :t:`else expression` has a :t:`block expression`, infer the
    :t:`block expression` with the :t:`expected type` of the :t:`if expression`

>     or :t:`if-let expression`.

Reviewer: ```suggestion
    or :t:`if let expression`.
```

  * :dp:`fls_Mcpwyvz47SoG`
    If the :t:`else expression` has an :t:`if expression`, infer that nested
    :t:`if expression` with the :t:`expected type` of the original
    :t:`if expression`, then :t:`unify` its :t:`type` with the :t:`type` of

>     the original :t:`if expression` or :t:`if-let expression`.

Reviewer: ```suggestion
    the original :t:`if expression` or :t:`if let expression`.
```

  * :dp:`fls_34AQ9g7xhdUj`

>     Otherwise, the :t:`else expression` has an :t:`if-let expression`. Infer
>     that nested :t:`if-let expression` with the :t:`expected type` of the
>     original :t:`if expression`, then :t:`unify` its :t:`type` with the
>     :t:`type` of the original :t:`if expression` or :t:`if-let expression`.

Reviewer: ```suggestion
    Otherwise, the :t:`else expression` has an :t:`if let expression`. Infer
    that nested :t:`if let expression` with the :t:`expected type` of the
    original :t:`if expression`, then :t:`unify` its :t:`type` with the
    :t:`type` of the original :t:`if expression` or :t:`if let expression`.
```

* :dp:`fls_4ZT35povCL04`
  A :t:`match expression` is inferred as follows:
  
  #. :dp:`fls_62OcWZaVN9hh`
     :t:`Unify` the :t:`[type]s` of the :t:`[pattern]s` of every :t:`match arm`,
     then infer the :t:`subject expression` with the :t:`expected type` set to
     the :t:`type` of the :t:`[pattern]s`.

  #. :dp:`fls_st9onPgDrc8y`
     Infer the :t:`[operand]s` of all :t:`[match arm guard]s` with
     :t:`expected type` :c:`bool`.

  #. :dp:`fls_F999gqcBfff9`
     Infer the :t:`match arm body` of every :t:`match arm` with the
     :t:`expected type` of the :t:`match expression`.

* :dp:`fls_Esa4ST7lLp8T`
  A :t:`for loop expression` is inferred by unifying the :t:`type` of its

>   :t:`pattern` with the :t:`type` ``<$T as core::iter::IntoIterator>::Item``,
>   where ``$T`` is the :t:`type` of the :t:`subject expression`, and then

Reviewer: ```suggestion
  :t:`pattern` with the :t:`type` ``<T as core::iter::IntoIterator>::Item``,
  where ``T`` is the :t:`type` of the :t:`subject expression`, and then
```

  inferring its :t:`loop body`.

* :dp:`fls_9GDElCkL1UbH`
  A :t:`while let loop expression` is inferred by unifying the :t:`type` of its
  :t:`subject let expression` with the :t:`type` of its :t:`pattern`, and then
  inferring its :t:`loop body`.

* :dp:`fls_0eATa6RtDNtA`
  An :t:`array expression` using an :t:`array element constructor` is inferred
  by attempting a :t:`least upper bound coercion` from each element :t:`type` to
  the :t:`expected type`. If no such :t:`type coercion` is possible, all element
  :t:`[type]s` are unified instead.

* :dp:`fls_q1JZZMxqWXCk`
  A :t:`negation expression` is inferred as follows:

  #. :dp:`fls_hH58ftCxBYzm`
     Determine the :t:`trait` corresponding to the operator according to the
     following table:

     .. list-table::

       * - :dp:`fls_aiSI99pbAYqT`
         - **Operator**
         - **Trait**
       * - :dp:`fls_zRdxowO4eDMN`
         - ``!``
         - :std:`core::ops::Not`
       * - :dp:`fls_IoceMi7HfqsK`
         - ``-``
         - :std:`core::ops::Neg`

  #. :dp:`fls_lDkPMB5UI58B`
     Infer the :t:`type` of the :t:`expression` to be the :t:`associated type`

>      ``<$T as $Trait>::Output``, where ``$T`` is the :t:`type` of the
>      :t:`operand`, and ``$Trait`` is the operator :t:`trait` determined from

Reviewer: ```suggestion
     ``<T as Trait>::Output``, where ``T`` is the :t:`type` of the
     :t:`operand`, and ``Trait`` is the operator :t:`trait` determined from
```

       * - :dp:`fls_NpwLzJJH9cGw`
         - ``%``
         - :std:`core::ops::Mod`
       * - :dp:`fls_56J8BlLOuvr4`
         - ``&``
         - :std:`core::ops::BitAnd`
       * - :dp:`fls_jK2pIVxOmtJ8`
         - ``|``
         - :std:`core::ops::BitOr`
       * - :dp:`fls_fjV22WcosNnt`
         - ``^``
         - :std:`core::ops::BitXor`
       * - :dp:`fls_h3OVuCdsKPhV`
         - ``<<``
         - :std:`core::ops::Shl`
       * - :dp:`fls_be2djziKJw3I`
         - ``>>``
         - :std:`core::ops::Shr`

  #. :dp:`fls_nHt0LVSiwTB3`
     If the :t:`expression` is a :t:`shift left expression` or a
     :t:`shift right expression`, and the :t:`expected type` is an
     :t:`integer type`, :t:`unify` the :t:`type` of the :t:`left operand` with
     the :t:`expected type`.

  #. :dp:`fls_sLCBZ3vG1AWs`
     If the :t:`expression` is neither a :t:`shift left expression` nor a
     :t:`shift right expression`, and the :t:`expected type` is a
     :t:`numeric type`, :t:`unify` the :t:`[type]s` of both :t:`[operand]s` with
     the :t:`expected type`.

  #. :dp:`fls_mCISAdm7sjRs`
     Infer the :t:`type` of the :t:`expression` to be the :t:`associated type`

>      ``<$L as $Trait<$R>>::Output``, where ``$L`` is the :t:`type` of the
>      :t:`left operand`, ``$Trait`` is the operator :t:`trait` determined from
>      the table above, and ``$R`` is the :t:`type` of the :t:`right operand`.

Reviewer: ```suggestion
     ``<L as Trait<R>>::Output``, where ``L`` is the :t:`type` of the
     :t:`left operand`, ``Trait`` is the operator :t:`trait` determined from
     the table above, and ``R`` is the :t:`type` of the :t:`right operand`.
```

         - :std:`core::ops::SubAssign`
       * - :dp:`fls_A36NBOl1FTCb`
         - ``*=``
         - :std:`core::ops::MulAssign`
       * - :dp:`fls_h3mmmIBR72kV`
         - ``/=``
         - :std:`core::ops::DivAssign`
       * - :dp:`fls_8edzBBIo7jF7`
         - ``%=``
         - :std:`core::ops::ModAssign`
       * - :dp:`fls_lUg26vFuSePP`
         - ``&=``
         - :std:`core::ops::BitAndAssign`
       * - :dp:`fls_21ay7EUUUmhx`
         - ``|=``
         - :std:`core::ops::BitOrAssign`
       * - :dp:`fls_8VgAhOgDOk0y`
         - ``^=``
         - :std:`core::ops::BitXorAssign`
       * - :dp:`fls_OVVY9CE0pGtJ`
         - ``<<=``
         - :std:`core::ops::ShlAssign`
       * - :dp:`fls_FojOvB6l3lAh`
         - ``>>=``
         - :std:`core::ops::ShrAssign`

  #. :dp:`fls_CVfHkJq1PixR`
     Introduce an :t:`obligation` ``$L: $Trait<$R>``, where ``$L`` is the
     :t:`type` of the :t:`assigned operand`, ``$Trait`` is the operator
     :t:`trait` determined from the table above, and ``$R`` is the :t:`type` of
     the :t:`modifying operand`.

  #. :dp:`fls_0RZ7w0YqmzE3`
     The :t:`type` of the :t:`expression` is the :t:`unit type`.

* :dp:`fls_YppNCEPMYqWJ`

>   A :t:`comparison expression` is inferred by introducing an :t:`obligation`
>   ``$L: PartialEq<$R>``, where ``$L`` is the :t:`type` of the :t:`left operand`,
>   and ``$R`` is the :t:`type` of the :t:`right operand`. The :t:`type` of the

Reviewer: ```suggestion
  A :t:`comparison expression` is inferred by introducing an :t:`obligation`
  ``L: PartialEq<R>``, where ``L`` is the :t:`type` of the :t:`left operand`,
  and ``R`` is the :t:`type` of the :t:`right operand`. The :t:`type` of the
```
